bionemo-recipes/recipes/llama_native_te_nvfsdp/dataloader.py

```diff
@@ -0,0 +1,276 @@
+from pathlib import Path
+import torch
+from torch.utils.data import Dataset, DataLoader, DistributedSampler
+from omegaconf import DictConfig
+
+from tokenizer import NucleotideASCIITokenizer
+
+
+class GenomicSequenceDataset(Dataset):
+    """
+    EVO2-style genomic dataset with proper tiling and randomization.
+    
+    Key features:
+    1. First tiling (pre-compute window mappings)
+    2. Then tokenization (ASCII nucleotide encoding)
+    3. Fixed positions + shuffled access (like EVO2)
+    """
+    
+    def __init__(
+        self,
+        database_path: str,
+        seq_length: int = 8192,
+        tokenizer: Optional[NucleotideASCIITokenizer] = None,
+        stride: int = 7992,  # EVO2 default: 200bp overlap
+        min_window_length: int = 1000,
+        seed: int = 42,
+        rc_augmentation: bool = False,
+    ):
+        """
+        Initialize genomic dataset with EVO2-style tiling.
+        
+        Args:
+            database_path: Path to SQLite database with sequences
+            seq_length: Window size (8192 for LLAMA3)
+            tokenizer: ASCII tokenizer for nucleotides
+            stride: Stride between windows (7992 = 8192-200 for EVO2 overlap)
+            min_window_length: Minimum effective window length
+            seed: Random seed for reproducible randomization
+            rc_augmentation: Apply reverse complement augmentation
+        """
+        self.database_path = Path(database_path)
+        self.seq_length = seq_length
+        self.stride = stride
+        self.min_window_length = min_window_length
+        self.seed = seed
+        self.rc_augmentation = rc_augmentation
+        
+        if tokenizer is None:
+            self.tokenizer = NucleotideASCIITokenizer()
+        else:
+            self.tokenizer = tokenizer
+            
+        # Initialize random state
+        self.rng = np.random.RandomState(seed)
+        random.seed(seed)
+        
+        # Step 1: Load sequences from database
+        self._load_sequences()
+        
+        # Step 2: First create window mappings (EVO2 style)
+        self._create_window_mappings()
+        
+        # Step 3: Randomize lookups - shuffle window indices
+        self._randomize_window_access()
+        
+        print(f"Dataset ready: {len(self.window_mappings):,} windows with EVO2-style tiling")
+    
+    def _load_sequences(self):
+        """Load sequence metadata from SQLite database."""
+        if not self.database_path.exists():
+            raise FileNotFoundError(f"Database not found: {self.database_path}")
+            
+        with sqlite3.connect(str(self.database_path)) as conn:
+            cursor = conn.cursor()
+            cursor.execute("SELECT contig_id, length FROM sequences ORDER BY contig_id")
+            self.sequences = cursor.fetchall()
+            
+        print(f"Loaded {len(self.sequences)} sequences from database")
+    
+    def _create_window_mappings(self):
```

Comment:
why does the `stride` parameter from hf tokenizers not do this?
https://huggingface.co/docs/transformers/en/main_classes/tokenizer